Sources/Fullscreen/FullscreenGallery/FullscreenGalleryViewController.swift

```diff
@@ -0,0 +1,472 @@
+        super.viewDidLoad()
+
+        view.backgroundColor = .clear
+
+        view.addSubview(backgroundView)
+        view.sendSubviewToBack(backgroundView)
+        backgroundView.fillInSuperview()
+
+        view.addSubview(captionLabel)
+        view.addSubview(dismissButton)
+        view.addSubview(previewView)
+        view.addGestureRecognizer(singleTapGestureRecognizer)
+
+        let initialPreviewViewVisibilityConstraint = previewViewVisible
+            ? previewViewVisibleConstraint
+            : previewViewHiddenConstraint
+
+        NSLayoutConstraint.activate([
+            captionLabel.centerXAnchor.constraint(equalTo: safeLayoutGuide.centerXAnchor),
+            captionLabel.widthAnchor.constraint(lessThanOrEqualTo: safeLayoutGuide.widthAnchor, constant: -(2 * CGFloat.mediumLargeSpacing)),
+            captionLabel.bottomAnchor.constraint(lessThanOrEqualTo: safeLayoutGuide.bottomAnchor, constant: -.mediumSpacing),
+            captionLabel.bottomAnchor.constraint(lessThanOrEqualTo: previewView.topAnchor, constant: -.mediumSpacing),
+
+            dismissButton.leadingAnchor.constraint(equalTo: safeLayoutGuide.leadingAnchor, constant: .mediumLargeSpacing),
+            dismissButton.topAnchor.constraint(equalTo: safeLayoutGuide.topAnchor, constant: .mediumLargeSpacing),
+            dismissButton.widthAnchor.constraint(equalToConstant: FullscreenGalleryViewController.dismissButtonSize),
+            dismissButton.heightAnchor.constraint(equalToConstant: FullscreenGalleryViewController.dismissButtonSize),
+
+            previewView.leadingAnchor.constraint(lessThanOrEqualTo: view.leadingAnchor),
+            previewView.trailingAnchor.constraint(lessThanOrEqualTo: view.trailingAnchor),
+            previewView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+            initialPreviewViewVisibilityConstraint
+        ])
+
+        view.layoutIfNeeded()
+
+        transitionToImage(atIndex: viewModel.selectedIndex, animated: false)
+    }
+
+    public override func viewDidLayoutSubviews() {
+        super.viewDidLayoutSubviews()
+
+        currentImageViewController()?.updateLayout(withPreviewViewVisible: previewViewVisible)
+
+        if !hasPerformedInitialPreviewScroll {
+            if let currentIndex = currentImageViewController()?.imageIndex {
+                previewView.layoutIfNeeded()
+                previewView.scrollToItem(atIndex: currentIndex, animated: false)
+                hasPerformedInitialPreviewScroll = true
+            }
+        }
+    }
+
+    public override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
+        super.viewWillTransition(to: size, with: coordinator)
+        previewView.superviewWillTransition(to: size)
+    }
+
+    // MARK: - View interactions
+
+    @objc private func dismissButtonTapped() {
+        galleryTransitioningController?.dismissVelocity = nil
+        dismiss(animated: true)
```

Comment:
Sounds like a good idea 👍 This would also make the call to `DeviceOrientation.lockPhone()` from `FINObjectViewController` happen in a more natural place 🙂 It currently happens in `FullscreenGalleryTransitionPresenterDelegate::fullscreenGalleryTransitionOutCompleted`, which only gets called when using the **optional** transition controller.